www/source/blog/2017-10-15-Configuration-Plans.html.md

```diff
@@ -0,0 +1,119 @@
+---
+title: Configuration Plans
+date: 2017-10-15
+author: Thom May
+tags: [plans, configuration, glossary]
+category: Packaging
+classes: body-article
+---
+
+One of the hardest problems in computer science is naming things. But one of the
+most important things for a healthy community is shared language. As the
+Habitat community grows, we'll try to identify patterns and practices
+that help us succeed, and give them names so that we can all talk about
+them.
+
+One pattern we use a lot is what we're calling a Configuration Plan. As
+we share our packages, we've been seeing that our colleagues and
+community members use those packages in ways that we could not have
+anticipated. So what to do? We could try and add templates to
+support all possible uses, but that will rapidly become untenable. What
+good packages should do instead is provide well documented entry points, so
+that consumers of our package can easily supply their own configuration
+in a new plan -- a Configuration Plan.
+
+## Building a Configuration Plan
+
+Let's have a look at an example. There are actually a fair number of
+live examples in the habitat repository - check out the [Builder API
+Proxy](https://github.com/habitat-sh/habitat/tree/master/components/builder-api-proxy), which is a Configuration Plan for nginx. But for now, let's build our own.
+
+I've created a tiny binary that simply reads a file and prints it out.
+You can [see the
+source](https://github.com/thommay/configuration_plan_demo), and you can
+try running it by entering a studio and running the demo.
+
+```bash
+$ hab studio enter
+(studio) $ hab svc start thom/configuration_plan_demo
+(studio) $ sup-log
+```
+
+You'll see the demo print out the configuration file name, and then the
+contents of the file.
+
+```
+configuration_plan_demo.default(O): Filename is: /hab/svc/configuration_plan/config/output.txt
+configuration_plan_demo.default(O): The original plan
+
+### Making a new plan
+Now, let's create a Configuration Plan that provides a new file for the
+demo. First, we'll create a new plan:
+
+```bash
+$ hab plan init -o <your_origin> configuration_plan
```

Comment:
What's the actual naming convention here when someone creates a new plan that wraps your plan and provides new configuration.

* Would you name it the same? Essentially you are using it in the same way. But I could see for some applications (looking at you *nginx*) that when given different configuration files completely change.

* Would you name it what it becomes? In this case if you were to give it a whole new configuration and it performs in a different capacity would you call it something different? I think when we were building some learning content around this we defaulted to call our configuration plan version of `core/httpd` the `learn-chef/webapp` as it now behaved in way that served our web content.

> An aside, we also did something which was probably an anti-pattern and packaged our website with the updated config files. I saw in the nginx blog post the concept of storing the site in a completely different package and installing it this webapp one